File: packages/provide-uterm-server/src/provide/uterm/server/app/posture.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""Unified security-posture self-report for the hosted terminal server.

``compute_security_posture`` collapses the ~8 independent security knobs the
server has accumulated into a single JSON-safe dict describing the *effective*
posture of a running configuration: which dev opt-outs are active, what bind
host is in use, and one boolean ``secure`` summary. It is a pure function (no
I/O) so it can be logged at startup and served from an auth-gated endpoint.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from provide.uterm.server.app.auth import _is_loopback_host

if TYPE_CHECKING:
    from provide.uterm.server.models import ServerConfig


def compute_security_posture(config: ServerConfig) -> dict[str, object]:
    """Return a JSON-safe dict describing the effective security posture.

    Pure function: reads only ``config`` and returns plain JSON types. Each
    entry in ``dev_opt_outs`` names an *active, security-weakening* opt-out.

    A knob counts as "weakening" only when it is actually relaxing a control:

      * ``auth.mode=dev_token`` — stub-IdP dev auth (startup-gated to loopback).
      * ``security.mode=dev`` — strips HSTS/CSP/X-Frame-Options (startup-gated
        to loopback unless acknowledged).
      * ``auth.webhook_idp_on_failure=viewer`` — fail-open to an anonymous
        viewer principal when the webhook IdP errors.
      * ``auth.header_mode_acknowledged`` — only weakening when paired with
        ``auth.mode=header`` (it acknowledges trusting X-Uterm-Role from
        callers); inert otherwise.
      * ``security.dev_mode_acknowledged`` — only weakening when paired with
        ``security.mode=dev`` (acknowledges relaxed headers on a routable
        bind); inert otherwise.
      * ``auth.allow_adhoc_browser_observers`` — lets non-admins observe
        unregistered (ad-hoc) workers.
      * ``security.block_private_connector_targets=false`` — connectors may
        reach internal/loopback/link-local hosts. Default-False is the tool's
        intended use, but in posture terms it is a relaxation worth surfacing.
    """
    bind_host = str(config.server.host).strip().lower()
    is_loopback = _is_loopback_host(bind_host)
    # ``dev_token`` collapses to ``jwt`` once setup_dev_idp runs at startup. The
    # declared mode is stamped on the AuthConfig by that path; prefer it so the
    # posture report surfaces the dev opt-out rather than the post-mutation jwt.
    auth_mode = str(getattr(config.auth, "_declared_auth_mode", None) or config.auth.mode).strip().lower()
    security_mode = str(config.security.mode).strip().lower()

    dev_opt_outs: list[str] = []
    warnings: list[str] = []

    if auth_mode == "dev_token":
        dev_opt_outs.append("auth.mode=dev_token")
        warnings.append("dev_token auth on a non-loopback bind is refused at startup")

    if security_mode == "dev":
        dev_opt_outs.append("security.mode=dev")
        warnings.append("security.mode=dev strips HSTS/CSP/X-Frame-Options")
        # dev_mode_acknowledged only weakens posture when it actually unlocks
        # the relaxed header set (i.e. security.mode=dev).
        if config.security.dev_mode_acknowledged:
            dev_opt_outs.append("security.dev_mode_acknowledged")

    if str(getattr(config.auth, "webhook_idp_on_failure", "deny")).strip().lower() == "viewer":
        dev_opt_outs.append("auth.webhook_idp_on_failure=viewer")
        warnings.append("anonymous-viewer IDP fallback is enabled (webhook_idp_on_failure=viewer)")

    # 1f: webhook IdP without response-signature verification → a forged response
    # can mint a principal. Only a weakening opt-out for the webhook provider.
    identity_provider = str(getattr(config.auth, "identity_provider", "local")).strip().lower()
    require_signed_response = bool(getattr(config.auth, "webhook_idp_require_signed_response", True))
    if identity_provider == "webhook" and not require_signed_response:
        dev_opt_outs.append("auth.webhook_idp_require_signed_response=false")
        warnings.append("webhook IdP responses are not signature-verified — a forged response can mint a principal")

    # header_mode_acknowledged only weakens posture in header auth mode.
    if auth_mode == "header" and config.auth.header_mode_acknowledged:
        dev_opt_outs.append("auth.header_mode_acknowledged")
        warnings.append("header auth trusts X-Uterm-Role headers from callers")

    if config.auth.allow_adhoc_browser_observers:
        dev_opt_outs.append("auth.allow_adhoc_browser_observers")
        warnings.append("non-admins may observe unregistered (ad-hoc) workers")

    if not config.security.block_private_connector_targets:
        dev_opt_outs.append("security.block_private_connector_targets=false (connectors may reach internal hosts)")

    # Compliance posture: the tamper-evident WORM audit chain. Defensive getattr
    # keeps the report working for embedder configs predating the ``audit`` field.
    audit_chain_enabled = bool(getattr(getattr(config, "audit", None), "chain_enabled", False))
    if not audit_chain_enabled:
        # Not a dev opt-out (it doesn't relax a control) — a compliance note: the
        # audit log is only a plain log stream, not integrity-chained.
        warnings.append(
            "audit log is not tamper-evident (audit.chain_enabled=false) — enable the WORM chain for compliance"
        )

    dev_opt_outs.sort()

    # ``secure`` is True iff this is a declared production deployment AND no
    # weakening opt-out is dangerous on this bind. A weakening opt-out is only
    # dangerous when the listener is remotely reachable: on a loopback bind the
    # deployment cannot be reached off-box, so loopback-only relaxations don't
    # demote the posture. Hence: production + (loopback OR no active opt-out).
    secure = config.environment == "production" and (is_loopback or not dev_opt_outs)

    return {
        "environment": str(config.environment),
        "bind_host": bind_host,
        "is_loopback": is_loopback,
        "auth_mode": auth_mode,
        "dev_opt_outs": dev_opt_outs,
        # 1f/1d: whether webhook-IDP response signing is required (real field,
        # secure-by-default True). Defensive getattr keeps the report JSON-safe
        # for any embedder config object that predates the field.
        "idp_signing_required": getattr(config.auth, "webhook_idp_require_signed_response", None),
        # Compliance: whether the tamper-evident WORM audit chain is enabled.
        "audit_chain_enabled": audit_chain_enabled,
        "warnings": warnings,
        "secure": secure,
    }

File: packages/provide-uterm-ts/src/serverconfig/posture.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * The server's security posture, as one report.
 *
 * Port of `provide.uterm.server.app.posture`. Eight or so independent knobs
 * collapsed into a single answer: which weakening opt-outs are actually
 * active, what the server is bound to, and one `secure` summary. Pure, so it
 * can be logged at startup and served from an auth-gated endpoint.
 */

/** The hosts that mean "cannot be reached off-box". */
const LOOPBACK_HOSTS: ReadonlySet<string> = new Set(["127.0.0.1", "localhost", "::1", "[::1]", "0:0:0:0:0:0:0:1"]);

/** Just the fields the report reads. */
export interface PostureConfig {
  environment: string;
  server: { host: string };
  auth: Record<string, unknown>;
  security: Record<string, unknown>;
  audit?: { chain_enabled?: boolean } | undefined;
}

/** What the report says. */
export interface SecurityPosture {
  environment: string;
  bind_host: string;
  is_loopback: boolean;
  auth_mode: string;
  /** Active, security-weakening opt-outs, sorted so the report reads the same every time. */
  dev_opt_outs: string[];
  idp_signing_required: boolean | null;
  idp_response_replay_protected: boolean;
  audit_chain_enabled: boolean;
  warnings: string[];
  secure: boolean;
}

/** Normalise a value a person wrote in a config file. */
function text(value: unknown, fallback = ""): string {
  return String(value ?? fallback)
    .trim()
    .toLowerCase();
}

/** Whether a bind address can be reached off-box. */
function isLoopbackHost(host: string): boolean {
  return LOOPBACK_HOSTS.has(host);
}

/**
 * Report the effective posture of a running configuration.
 *
 * An opt-out counts only where it is actually relaxing something. Two of the
 * knobs are acknowledgements: they weaken nothing on their own and matter
 * only paired with the mode they unlock. Listing them regardless would report
 * a posture worse than the deployment has, and a report that cries wolf stops
 * being read.
 */
export function computeSecurityPosture(config: PostureConfig): SecurityPosture {
  const bindHost = text(config.server.host);
  const isLoopback = isLoopbackHost(bindHost);
  // `dev_token` collapses to `jwt` once the dev IdP is set up at startup. The
  // declared mode is preferred so the report surfaces the opt-out rather than
  // the mode it turned into.
  const authMode = text(config.auth._declared_auth_mode ?? config.auth.mode);
  const securityMode = text(config.security.mode);

  const devOptOuts: string[] = [];
  const warnings: string[] = [];

  if (authMode === "dev_token") {
    devOptOuts.push("auth.mode=dev_token");
    warnings.push("dev_token auth on a non-loopback bind is refused at startup");
  }

  if (securityMode === "dev") {
    devOptOuts.push("security.mode=dev");
    warnings.push("security.mode=dev strips HSTS/CSP/X-Frame-Options");
    // Only weakening where it actually unlocks the relaxed header set.
    if (config.security.dev_mode_acknowledged === true) {
      devOptOuts.push("security.dev_mode_acknowledged");
    }
  }

  if (text(config.auth.webhook_idp_on_failure, "deny") === "viewer") {
    devOptOuts.push("auth.webhook_idp_on_failure=viewer");
    warnings.push("anonymous-viewer IDP fallback is enabled (webhook_idp_on_failure=viewer)");
  }

  const identityProvider = text(config.auth.identity_provider, "local");
  const requireSignedResponse = config.auth.webhook_idp_require_signed_response !== false;
  if (identityProvider === "webhook" && !requireSignedResponse) {
    devOptOuts.push("auth.webhook_idp_require_signed_response=false");
    warnings.push("webhook IdP responses are not signature-verified — a forged response can mint a principal");
  }

  // The always-on per-instance replay cache blocks verbatim replay within one
  // process but not across nodes; the nonce binding closes that gap.
  const requireResponseNonce = config.auth.webhook_idp_require_response_nonce === true;
  if (identityProvider === "webhook" && !requireResponseNonce) {
    warnings.push(
      "webhook IdP response replay is blocked only by a per-instance cache (not shared across nodes) — " +
        "enable auth.webhook_idp_require_response_nonce for HA / strict request-binding",
    );
  }

  if (authMode === "header" && config.auth.header_mode_acknowledged === true) {
    devOptOuts.push("auth.header_mode_acknowledged");
    warnings.push("header auth trusts X-Uterm-Role headers from callers");
  }

  if (config.auth.allow_adhoc_browser_observers === true) {
    devOptOuts.push("auth.allow_adhoc_browser_observers");
    warnings.push("non-admins may observe unregistered (ad-hoc) workers");
  }

  if (config.security.block_private_connector_targets !== true) {
    devOptOuts.push("security.block_private_connector_targets=false (connectors may reach internal hosts)");
  }

  const auditChainEnabled = config.audit?.chain_enabled === true;
  if (!auditChainEnabled) {
    // A compliance note rather than an opt-out: it does not relax an existing
    // control, it is the absence of a stronger one.
    warnings.push("audit log is not tamper-evident (audit.chain_enabled=false) — enable the WORM chain for compliance");
  }

  devOptOuts.sort();

  return {
    environment: String(config.environment),
    bind_host: bindHost,
    is_loopback: isLoopback,
    auth_mode: authMode,
    dev_opt_outs: devOptOuts,
    idp_signing_required:
      config.auth.webhook_idp_require_signed_response === undefined
        ? null
        : config.auth.webhook_idp_require_signed_response === true,
    // Unconditional: the per-instance cache is always on.
    idp_response_replay_protected: true,
    audit_chain_enabled: auditChainEnabled,
    warnings,
    // Production *and* (loopback or nothing relaxed). A weakening opt-out is
    // only dangerous where the listener is remotely reachable, so a
    // loopback-only relaxation does not demote the posture.
    secure: config.environment === "production" && (isLoopback || devOptOuts.length === 0),
  };
}
